### PR TITLE
Pattern support

### PR DESCRIPTION
**Pull request by garrett-wade**
## Adding support for custom comment pattern specification

This is initial enhancement to support the following issue https://github.com/readmeio/oas/issues/603.

This enhancement:

- added an optional `pattern` argument command-line argument with help text
- added pattern attribute and get/set methods to `Options` object
- added `loadPattern` method to `Loader` to load json file that contains specification of pattern
- added section in `swaggerInline` in `index` file to handle adding the loaded pattern file to the options `variable`
- ensured backwards compatibility if no pattern is specified
- expanded test coverage to cover all changes

### Testing

Refer to the `can extract comments from custom pattern and custom file` test in the `extractor.test.js` file to see a detailed test of the additional functionality.

### CODE REVIEW

File: src/cli.js [hunk 1]
```diff
@@ -9,7 +9,8 @@ module.exports = function (args) {
     .usage('[options] <inputGlobs ...>')
     .option('-b, --base [path]', 'A base OpenAPI or Swagger definition to build off of.')
     .option('-f, --format [format]', 'Output format (.json or .yaml).')
-    .option('-s, --scope [scope]', 'API scope to constrain against');
+    .option('-s, --scope [scope]', 'API scope to constrain against')
+    .option('-p, -- pattern [pattern]', 'Custom comment pattern supported by the multilang-extract-comments library in the form of a JSON file');
```
Comment (erunion) [inline src/cli.js:13]:
```suggestion
    .option('-p, --pattern [pattern]', 'Custom comment pattern supported by the multilang-extract-comments library in the form of a JSON file');
```

File: __tests__/extractor.test.js [hunk 2]
```diff
@@ -17,6 +17,24 @@ describe('Extractor', () => {
...
+
+      it('can extract comments from custom pattern and custom file', () => {
+        const code = fs.readFileSync(`${__dirname}/__fixtures__/code/apex.cls`, 'utf-8');
+        const customPattern = JSON.parse((fs.readFileSync(`${__dirname}/__fixtures__/patterns/pattern-valid.json`)));
+        const comments = Extractor.extractComments(code, {
+          filename: 'apex.cls',
+          pattern: customPattern
+        });
+
+
```
Comment (erunion) [inline __tests__/extractor.test.js:37]:
```suggestion
```

File: __tests__/extractor.test.js [hunk 3]
```diff
@@ -17,6 +17,24 @@ describe('Extractor', () => {
...
+        Object.keys(comments).forEach((key, index) => {
+          const comment = comments[key];
+          expect(comment.info.type).toBe(expectations[index].type);
+          expect(comment.content).toContain(expectations[index].content);
+        });
+      });
+
+      it('can extract comments from custom pattern and custom file', () => {
+        const code = fs.readFileSync(`${__dirname}/__fixtures__/code/apex.cls`, 'utf-8');
+        const customPattern = JSON.parse((fs.readFileSync(`${__dirname}/__fixtures__/patterns/pattern-valid.json`)));
```
Comment (erunion) [inline __tests__/extractor.test.js:31]:
```suggestion
        const customPattern = JSON.parse(fs.readFileSync(`${__dirname}/__fixtures__/patterns/pattern-valid.json`));
```

File: src/extractor.js [hunk 4]
```diff
@@ -48,8 +49,9 @@ function buildSchema(schema, yamlLines) {
 
 class Extractor {
   static extractEndpointsFromCode(code, options) {
+    
     const comments = this.extractComments(code, options);
-
+
```
Comment (erunion) [inline src/extractor.js:54]:
```suggestion

```

File: src/extractor.js [hunk 5]
```diff
@@ -1,5 +1,6 @@
 const extractComments = require('multilang-extract-comments');
 const jsYaml = require('js-yaml');
+const { option } = require('commander');
```
Comment (erunion) [inline src/extractor.js:3]: Seems this import is unused?

File: src/index.js [hunk 6]
```diff
@@ -4,6 +4,7 @@ const path = require('path');
 const Loader = require('./loader');
 const Extractor = require('./extractor');
 const Options = require('./options');
+const { option } = require('commander');
```
Comment (erunion) [inline src/index.js:7]: Seems this import is unused?

File: src/loader.js [hunk 7]
```diff
@@ -48,6 +48,15 @@ class Loader {
     return Loader.LOADER_METHODS[extname](filepath, options);
   }
 
+  static loadPattern(filepath) {
+    return Loader.loadFile(filepath).then(pattern => {
+      return JSON.parse(pattern);
+    }).catch((err) => {
+      // Return null if we have any problems, potetial for more rubust error handling.
```
Comment (erunion) [inline src/loader.js:55]:
```suggestion
      // Return null if we have any problems, potential for more rubust error handling.
```

**Review by erunion — CHANGES_REQUESTED** [review]
Handful of nits, but overall this looks pretty solid! Thanks for picking this up.

Can you also re-run Prettier on your changes? You can run it with `npx prettier --list-different --write "./**/**.{js}"`

**Comment by garrett-wade** [comment]
@erunion Receiving following error when running prettier: `[error] No files matching the pattern were found: "./**/**.{js}".` Any thoughts here?

**Comment by garrett-wade** [comment]
@erunion was able to run

File: __tests__/loader.test.js [hunk 8]
```diff
@@ -67,6 +67,35 @@ describe('Loader', () => {
...
+  describe('#loadPattern', () => {
+    it('returns a pattern object', () => {
+      const patternPath = '${__dirname}/__fixtures__/patterns/pattern-valid.json';
+      return Loader.loadPattern(patternPath).then(pattern => {
+        expect(pattern).toBeDefined();
+      });
+    });
+
+    it('returns null if the file is invalid json', () => {
+      const patternPath = '${__dirname}/__fixtures__/patterns/pattern-valid.json';
```
Comment (erunion) [inline __tests__/loader.test.js:79]: The test workflows aren't running on this PR so I haven't had a chance to verify tests locally yet but it seems that this one should load the invalid JSON file instead.

```suggestion
      const patternPath = '${__dirname}/__fixtures__/patterns/pattern-invalid.json';
```

File: package.json [hunk 9]
```diff
@@ -46,7 +46,7 @@
     "@readme/eslint-config": "^8.0.2",
     "eslint": "^8.2.0",
     "jest": "^27.0.3",
-    "prettier": "^2.0.1"
+    "prettier": "^2.5.1"
```
Comment (erunion) [inline package.json:49]: Can you revert this change and the `package-lock` changes? Looks like you're using npm@6 but we use npm@7. We can handle upgrading `prettier` out of scope of your work here.

**Review by erunion — CHANGES_REQUESTED** [review]

File: __tests__/loader.test.js [hunk 10]
```diff
@@ -67,6 +67,35 @@ describe('Loader', () => {
     });
   });
 
+  describe('#loadPattern', () => {
+    it('returns a pattern object', () => {
+      const patternPath = '${__dirname}/__fixtures__/patterns/pattern-valid.json';
+      return Loader.loadPattern(patternPath).then(pattern => {
+        expect(pattern).toBeDefined();
+      });
+    });
```
Comment (erunion) [inline __tests__/loader.test.js:76]: Running these tests locally I noticed that this isn't actually returning what it should be because the `patternPath` isn't a template string. This'll fix it and bulk up the code coverage a touch:

```suggestion
    it('returns a pattern object', async () => {
      const patternPath = `${__dirname}/__fixtures__/patterns/pattern-valid.json`;
      await expect(Loader.loadPattern(patternPath)).resolves.toStrictEqual({
        multiLineComment: [
          {
            end: '*/',
            middle: '',
            start: '/*',
          },
        ],
        name: 'Apex',
        nameMatchers: ['.cls'],
        singleLineComment: [
          {
            start: '//',
          },
        ],
      });
    });
```

File: __tests__/loader.test.js [hunk 11]
```diff
@@ -67,6 +67,35 @@ describe('Loader', () => {
...
+
+    it('returns null if the file is invalid json', () => {
+      const patternPath = '${__dirname}/__fixtures__/patterns/pattern-invalid.json';
+      return Loader.loadPattern(patternPath).then(pattern => {
+        expect(pattern).toBeNull();
+      });
+    });
```
Comment (erunion) [inline __tests__/loader.test.js:83]:
```suggestion
    it('returns null if the file is invalid json', async () => {
      const patternPath = `${__dirname}/__fixtures__/patterns/pattern-invalid.json`;
      await expect(Loader.loadPattern(patternPath)).resolves.toBeNull();
    });
```

File: __tests__/loader.test.js [hunk 12]
```diff
@@ -67,6 +67,35 @@ describe('Loader', () => {
...
+
+    it('returns null if the file does not exist', () => {
+      const patternPath = '${__dirname}/__fixtures__/patterns/pattern-that-does-not-exist.json';
+      return Loader.loadPattern(patternPath).then(pattern => {
+        expect(pattern).toBeNull();
+      });
+    });
```
Comment (erunion) [inline __tests__/loader.test.js:90]:
```suggestion
    it('returns null if the file does not exist', async () => {
      const patternPath = `${__dirname}/__fixtures__/patterns/pattern-that-does-not-exist.json`;
      await expect(Loader.loadPattern(patternPath)).resolves.toBeNull();
    });
```

File: src/extractor.js [hunk 13]
```diff
@@ -1,6 +1,5 @@
 const extractComments = require('multilang-extract-comments');
 const jsYaml = require('js-yaml');
-
```
Comment (erunion) [inline src/extractor.js:3]: Can you add this line break back in? Our coding standards wants it to be there:

Screen Shot 2022-03-01 at 10 25 00 PM

(You can see these errors with `npm run lint`)

File: src/index.js [hunk 14]
```diff
@@ -90,70 +90,82 @@ function swaggerInline(globPatterns, providedOptions) {
...
-                if (['.json', '.md', '.txt'].includes(path.extname(fileInfo.fileName))) {
-                  return Promise.resolve();
-                } else if (/Cannot find language definition/.test(err.message)) {
-                  return Promise.resolve();
+      return Loader.loadPattern(options.getPattern()).then(pattern => {
+        if (pattern) {
+          options.setPattern(pattern);
+        } else {
+          // if there is no valid pattern specifed reset the pattern attribute to it's default to prevent error.
+          options.setPattern(Options.DEFAULTS['pattern']);
```
Comment (erunion) [inline src/index.js:99]:
```suggestion
          options.setPattern(Options.DEFAULTS.pattern);
```

File: src/loader.js [hunk 15]
```diff
@@ -48,6 +48,17 @@ class Loader {
     return Loader.LOADER_METHODS[extname](filepath, options);
   }
 
+  static loadPattern(filepath) {
+    return Loader.loadFile(filepath)
+      .then(pattern => {
+        return JSON.parse(pattern);
+      })
+      .catch(err => {
```
Comment (erunion) [inline src/loader.js:56]: Our coding standards prefers unused variables not to be around:

```suggestion
      .catch(() => {
```

**Review by erunion — CHANGES_REQUESTED** [review]
Sorry, I finally had a chance to run this locally and a found a couple things. These should be the last of the changes! Thanks again for this.

**Comment by garrett-wade** [comment]
@erunion thanks for all of the feedback! This is my first time contributing to an open source library as well as extending a node.js library in general. Thank you for taking the time to make the suggestions, learned a lot!

**Review by erunion — APPROVED** [review]
Thanks again for this!

**Comment by erunion** [comment]
Published to https://github.com/readmeio/swagger-inline/releases/tag/5.1.0
